src/backend/marsha/core/views.py

```diff
@@ -81,72 +81,175 @@ def exception_handler(exc, context):
     frontend loads, frontend navigation takes over.
     """
 
+    frontend_name = "Site"
     template_name = "core/site.html"
     waffle_switch = "site"
 
+    def _get_app_data(self):
+        """Adds a user JWT to the default app data."""
+        app_data = super()._get_app_data()
+        app_data["jwt"] = str(UserAccessToken.for_user(self.request.user))
+        return app_data
+
     def get_context_data(self, **kwargs):
         """Build the context necessary to run the frontend app for the site."""
-        jwt_token = UserAccessToken.for_user(self.request.user)
+        return self._build_context_data(self._get_app_data())
 
-        app_data = _get_base_app_data()
-        app_data.update(
-            {
-                "frontend": "Site",
-                "jwt": str(jwt_token),
-            }
-        )
-
-        return {
-            "app_data": json.dumps(app_data),
-            "external_javascript_scripts": settings.EXTERNAL_JAVASCRIPT_SCRIPTS,
-            "static_base_url": f"{settings.STATIC_URL}js/build/",
-        }
 
+class ResourceException(Exception):
+    """Wrapper to normalize exceptions caught in views."""
 
-@method_decorator(csrf_exempt, name="dispatch")
-@method_decorator(xframe_options_exempt, name="dispatch")
-class BaseLTIView(ABC, TemplateResponseMixin, View):
-    """Base LTI view called to serve a resource on an LTI context
-    and how to use it by the front application.
 
-    It is designed to work as a React single page application.
+class BaseResourceView(TemplateResponseMixin, MarshaViewMixin, View):
+    """
+    Base view to provide common methods for view which concerns resource.
 
+    This does not inherit from `TemplateView` as the inheriting views may provide only POST method.
     """
 
+    frontend_name = "LTI"  # even for public resource
     template_name = "core/resource.html"
 
+    def _app_data_exception(self, error: BaseException):
+        """
+        Build app data in a context of failure (object not found, LTI authentication failed...)
+
+        This is called anytime a `ResourceException` is raised during
+        the `get_context_data` process.
+
+        Parameters
+        ----------
+        error : BaseException
+            The original exception raised.
+
+        Returns
+        -------
+        The `app data` dict.
+        """
+        logger.warning(str(error))
+        app_data = self._get_base_app_data()
+        app_data["state"] = "error"
+        return app_data
+
+    def _build_context_exception(self, error: BaseException):
+        """
+        Build the view context (for a `render_to_response`)
+        when a `ResourceException` is raised.
+
+        Parameters
+        ----------
+        error : BaseException
+            The original exception raised.
+
+        Returns
+        -------
+        The view's context dict.
+        """
+        return self._build_context_data(self._app_data_exception(error))
+
+    def _init_context(self):
```

Comment:
As we are already using it in this module, what about using `ABC` for this class and declare this method as abstract ?

Reply:
This method may stay "empty", we do not always need to override it. For instance, the `BaseView` does not add initialization here. Actually, it is only used for LTI views for now (to verify request).

Reply:
:+1: